pallets/pips/src/lib.rs

```diff
@@ -1342,29 +1355,72 @@ impl<T: Trait> Module<T> {
         Self::prune_data(did, id, new_state, Self::prune_historical_pips());
     }
 
-    fn schedule_pip_for_execution(did: IdentityId, id: PipId) {
-        // Set the default enactment period and move it to `Scheduled`
-        let curr_block_number = <system::Module<T>>::block_number();
-        let executed_at = curr_block_number + Self::default_enactment_period();
-
+    /// Adds a PIP execution call to the PIP execution schedule.
+    // TODO: the `maybe_at` argument is only required until we upgrade to a version of Substrate
+    // containing `schedule::Named::reschedule_named`.
+    fn schedule_pip_for_execution(did: IdentityId, id: PipId, maybe_at: Option<T::BlockNumber>) {
+        let at = maybe_at.unwrap_or_else(|| {
+            let period = Self::default_enactment_period();
+            let corrected_period = if period > Zero::zero() {
+                period
+            } else {
+                One::one()
+            };
+            <system::Module<T>>::block_number() + corrected_period
+        });
         Self::update_proposal_state(did, id, ProposalState::Scheduled);
-        <PipToSchedule<T>>::insert(id, executed_at);
-        <ExecutionSchedule<T>>::append(executed_at, id);
-        let event = RawEvent::ExecutionScheduled(did, id, Zero::zero(), executed_at);
+        <PipToSchedule<T>>::insert(id, at);
+
+        let call = Call::<T>::execute_scheduled_pip(id).into();
+        let event = if let Err(_) = T::Scheduler::schedule_named(
+            Self::pip_execution_name(id),
+            DispatchTime::At(at),
+            None,
+            HARD_DEADLINE,
+            RawOrigin::Root.into(),
+            call,
+        ) {
+            RawEvent::ExecutionSchedulingFailed(did, id, at)
+        } else {
+            RawEvent::ExecutionScheduled(did, id, Zero::zero(), at)
+        };
+        Self::deposit_event(event);
+    }
+
+    /// Adds a PIP expiry call to the PIP expiry schedule.
+    fn schedule_pip_for_expiry(id: PipId, at: T::BlockNumber) {
+        let did = GC_DID;
+        let call = Call::<T>::expire_scheduled_pip(did, id).into();
+        let event = if let Err(_) = T::Scheduler::schedule_named(
+            Self::pip_expiry_name(id),
+            DispatchTime::At(at),
+            None,
+            HARD_DEADLINE,
+            RawOrigin::Root.into(),
+            call,
+        ) {
+            RawEvent::ExpirySchedulingFailed(did, id, at)
+        } else {
+            RawEvent::ExpiryScheduled(did, id, at)
+        };
```

Comment:
Ignore this suggestion if you'd rather the current one.
```suggestion
        let event = match T::Scheduler::schedule_named(
            Self::pip_expiry_name(id),
            DispatchTime::At(at),
            None,
            HARD_DEADLINE,
            RawOrigin::Root.into(),
            call,
        ) {
            Err(_) => RawEvent::ExpirySchedulingFailed(did, id, at),
            Ok(_) => RawEvent::ExpiryScheduled(did, id, at),
        };
```